### PR TITLE
Path parent option

### PR DESCRIPTION
**Pull request by rgrinberg**
Calling Path.parent is a bit too dangerous as it may cause infinite loops in conjunction with recursion. Path.parent forces to think about the root case, and we still have Path.parent_exn for when we don't care.

### CODE REVIEW

File: src/stdune/path.ml [hunk 1]
```diff
@@ -340,10 +340,22 @@ let basename t =
...
+    let parent = Filename.dirname t in
+    if parent = t then
+      None
+    else
+      Some parent
+
+let parent_exn t =
+  match parent t with
+  | Some p -> p
+  | None -> Exn.code_error "Path.parent:exn t is root"
```
Comment (diml) [inline src/stdune/path.ml]: s/:/_/

**Comment by rgrinberg** [comment]
Typo is fixed.

**Comment by rgrinberg** [comment]
@diml I also added a commit to only have exception names in the failing expect tests rather than whole stack traces. The stack traces are too sensitive to arbitrary changes b/c they include line numbers.

**Comment by diml** [comment]
Ok, makes sense. I suppose we could also disable backtraces entirely in tests.

**Pull request merged**